nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/NiFiProvenanceEventAnalyzerFactory.java

```diff
@@ -0,0 +1,105 @@
+    private static final Map<ProvenanceEventType, NiFiProvenanceEventAnalyzer> analyzersForProvenanceEventType = new ConcurrentHashMap<>();
+    private static boolean loaded = false;
+
+    private static void loadAnalyzers() {
+        logger.debug("Loading NiFiProvenanceEventAnalyzer ...");
+        final ServiceLoader<NiFiProvenanceEventAnalyzer> serviceLoader
+                = ServiceLoader.load(NiFiProvenanceEventAnalyzer.class);
+        serviceLoader.forEach(analyzer -> {
+            addAnalyzer(analyzer.targetComponentTypePattern(), analyzersForComponentType, analyzer);
+            addAnalyzer(analyzer.targetTransitUriPattern(), analyzersForTransitUri, analyzer);
+            final ProvenanceEventType eventType = analyzer.targetProvenanceEventType();
+            if (eventType != null) {
+                if (analyzersForProvenanceEventType.containsKey(eventType)) {
+                    logger.warn("Fo ProvenanceEventType {}, an Analyzer {} is already assigned." +
+                            " Only one analyzer for a type can be registered. Ignoring {}",
+                            eventType, analyzersForProvenanceEventType.get(eventType), analyzer);
+                }
+                analyzersForProvenanceEventType.put(eventType, analyzer);
+            }
+        });
+        logger.info("Loaded NiFiProvenanceEventAnalyzers: componentTypes={}, transitUris={}", analyzersForComponentType, analyzersForTransitUri);
+    }
+
+    private static void addAnalyzer(String patternStr, Map<Pattern, NiFiProvenanceEventAnalyzer> toAdd,
+                                    NiFiProvenanceEventAnalyzer analyzer) {
+        if (patternStr != null && !patternStr.isEmpty()) {
+            Pattern pattern = Pattern.compile(patternStr.trim());
+            toAdd.put(pattern, analyzer);
+        }
+    }
+
+    /**
+     * Find and retrieve NiFiProvenanceEventAnalyzer implementation for the specified targets.
+     * Pattern matching is performed by following order, and the one found at first is returned:
+     * <ol>
+     * <li>Component type name. Use an analyzer supporting the Component type with its {@link NiFiProvenanceEventAnalyzer#targetProvenanceEventType()}.
+     * <li>TransitUri. Use an analyzer supporting the TransitUri with its {@link NiFiProvenanceEventAnalyzer#targetTransitUriPattern()}.
+     * <li>Provenance Event Type. Use an analyzer supporting the Provenance Event Type with its {@link NiFiProvenanceEventAnalyzer#targetProvenanceEventType()}.
+     * </ol>
+     * @param typeName NiFi component type name.
+     * @param transitUri Transit URI.
+     * @param eventType Provenance event type.
+     * @return Instance of NiFiProvenanceEventAnalyzer if one is found for the specified className, otherwise null.
+     */
+    public static NiFiProvenanceEventAnalyzer getAnalyzer(String typeName, String transitUri, ProvenanceEventType eventType) {
+
+        if (!loaded) {
```

Comment:
This is not thread-safe. This is an instance of Double-checked locking (https://en.wikipedia.org/wiki/Double-checked_locking)

Reply:
Thank you! I was referring an ancient blog post to implement this logic.. I will update it to use static lazy initializer.

Reply:
Specifically, "Initialization-on-demand holder idiom".
https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom